File: by-kotlin/kt-boj/solutions/10451-순열 사이클.kt
import java.util.*

fun main() {
    val sc = Scanner(System.`in`)
    val tc = sc.nextInt()

    for (i in 1..tc) {
        val n = sc.nextInt()
        val visited = mutableListOf<Boolean>()
        val adjacents = mutableListOf<Int>(0)
        for (j in 0..n) {
            visited.add(false)
        }
        for (j in 1..n) {
            adjacents.add(sc.nextInt())
        }

        var result = 0
        for (k in 1..n) {
            result += dfs(visited, adjacents, k)
        }
        println(result)
    }

}

fun dfs(visited: MutableList<Boolean>, adjacents: MutableList<Int>, num: Int): Int {
    if (adjacents[num] == num) {
        return 1
    }

    if (visited[num]) {
        return 0
    }

    visited[num] = true
    if (!visited[adjacents[num]]) {
        dfs(visited, adjacents, adjacents[num])
    }
    return 1
}

File: by-python/boj/10451-순열 사이클.py
tc = int(input())


def dfs(num):
    if adj[num] == num:
        visited[num] = True
        return 1

    if visited[num]:
        return 0

    visited[num] = True
    if not visited[adj[num]]:
        dfs(adj[num])

    return 1


for _ in range(tc):
    n = int(input())
    input_array = list(map(int, input().split()))

    adj = [[] for _ in range(n + 1)]
    visited = [False] * (n + 1)

    for i in range(1, n + 1):
        adj[i] = input_array[i - 1]

    result = 0
    for i in range(1, n + 1):
        result += dfs(i)

    print(result)
